src/utility.ts

```diff
@@ -20,3 +21,61 @@ export function signXmlResponse(samlMessage: string, options: SamlSigningOptions
     options
   );
 }
+
+export function assertObjectAndNotEmpty(value: any, error?: string): any {
+  if (typeof value !== "object" || JSON.stringify(value) === JSON.stringify({})) {
+    throw new TypeError(error ?? "value does not exist");
+  } else {
+    return value;
+  }
+}
+
+export function validateSAMLExtensionsElement(jsonSAMLExtensionsElement: any): boolean {
+  jsonSAMLExtensionsElement = assertObjectAndNotEmpty(
+    jsonSAMLExtensionsElement,
+    `samlExtensions Element value should be object and not empty`
+  );
+  let result = true;
+  for (const subElementKey in jsonSAMLExtensionsElement) {
+    if (!validateXMLNamespace(subElementKey, jsonSAMLExtensionsElement[subElementKey])) {
+      result = false;
+      break;
+    }
+  }
+  return result;
+}
+
+/**
+ * Validate the XMLBuilder input JSON, wether it has proper namespece or not
+ * @param jsonXMLElementKey - Key of the element, need key to understand its namespace
+ * @param jsonXMLElement - Just JSON which we want to pass to xmlbuilder to make xml
+ * @returns boolean | JSONXMLSchemaError
+ */
+export function validateXMLNamespace(
```

Comment:
IMHO reinventing/reimplementing XML validation does more harm than good.

This function seems to check only that top level element has namespace defined but if top level element does not define default namespace then this function should check that child elements / subtrees have namespace defined etc.

furthermore sometimes namespace prefix might already be defined at enclosing element (i.e. at AuthnRequest element level).

So at the end of the day this partial validation implementation just adds complexity but fails to catch all namespace issues (which it claims to validate). I repeat myself by saying that if nodejs ecosystem does not have pure javascript xml validation implementation which could be applied to whole AuthnRequest and if `node-gyp` based libraries are out of question then it would be easier from long term maintainability point of view to just let `node-saml` library users to debug their xml extension related issues with IdP admins (of course most of the developers would use e.g. online validation tools to debug whats wrong with their authrequests).

Reply:
This is a valid analysis. The best analysis of valid XML from our point of view, would be to simply check that the entire XML payload is valid (can be parsed by our XML tools), and that after it is parsed it contains an `<extensions>` element in the correct location.

While I don't want to be particular, it seems this would be as simple as trying to convert the JSON to XML and making sure that process doesn't fail. Then, using the XPath tools we already have to check to make sure that the `<extensions>` element is there with some expected attributes or children. This might not be possible during object construction, but we can certainly do it at run-time before we send the XML to anyone else.

If I'm looking at this from the wrong direction, perhaps we should just remove all validation and create a future work item to explore the entirety of XML validation. Honestly, any tool can be modified to present bad XML to an IdP. I'm just looking to alert the user of `node-saml` that they're doing something that isn't right so they can catch it and not have to rely on the IdP to tell them. But, maybe that is too much to ask or the wrong approach.

Reply:
>Then, using the XPath tools we already have to check to make sure that the <extensions> element is there with some expected attributes or children. This might not be possible during object construction, but we can certainly do it at run-time before we send the XML to anyone else.

@cjbarth you are aware that `ExtensionsType` allows anything?
```xml
    <complexType name="ExtensionsType">
        <sequence>
            <any namespace="##other" processContents="lax" maxOccurs="unbounded"/>
        </sequence>
    </complexType>
```
source: https://docs.oasis-open.org/security/saml/v2.0/saml-schema-protocol-2.0.xsd

I.e. from saml protocol layer point of view (and from protocol layer program blocks point of view such as `node-saml` or `passport-saml`) there aren't any "expected attributes or children" which existence/awailability `node-saml` could check on its own and on behalf of developer who is using `node-saml`.

Reply:
I understand that, but we can check to make sure that the bit of XML provided, in JSON form, is well-formed and has the correct opening tag...

Honestly, I feel like this is causing more problems than it is worth. I know debugging the XML coming out of `node-saml` is painful, so I wanted to make sure that we are doing due-diligence with the data we're working with. Perhaps we are better off with no validation what-so-ever, especially in light of https://github.com/node-saml/passport-saml/pull/551#issuecomment-894407649 and figuring out some other way to help the debugging experience along.

I'll adjust my opinion on this matter calling for a simple property to taken in the value and leaving it at that if @markstos agrees. The tests would then simply be making sure that the provided value gets injected in the outgoing XML at the correction location. Documentation would fill in the rest.

Reply:
I'll defer back to @cjbarth. Simple sounds good to me.

Reply:
In that case, lets not validate the incoming value for anything but being a valid object per your original idea @kdhttps . Sorry for the back-and-forth on this. @srd90 and @markstos provided valuable feedback and the iterative process sometimes works out this way.

Reply:
@cjbarth removed validation and test cases

Reply:
@cjbarth please review again and let me know if I miss anything